resources/page/permalinks.go

```diff
@@ -139,7 +144,12 @@ func init() {
 	}
 }
 
-func (l PermalinkExpander) parsePattern(pattern string) (func(Page) (string, error), error) {
+func (l PermalinkExpander) getOrParsePattern(pattern string) (func(Page) (string, error), error) {
+	expander, ok := l.patternCache.Get(pattern)
```

Comment:
This is nit picking, but the cache instance has a `GetOrCreate` method, so I suggest you make it

```go
return l.patternCache.GetOrCreate(pattern, parsePattern)
```

Reply:
I didn't like GetOrCreate because it needs to handle errors, but I can do something like this if it's what you're looking for:

```
	var err error
	expander := l.patternCache.GetOrCreate(pattern, func() func(Page) (string, error) {
		if !l.validate(pattern) {
			err = &permalinkExpandError{pattern: pattern, err: errPermalinkIllFormed}
			return nil
		}
		// ...
	})
	return expander, err
```

Reply:
GetOrCreate is used 2 other places, just change the signature to make it return an error.

Reply:
As I said, 

>GetOrCreate is used 2 other places, just change the signature to make it return an error.

But it's not important, I can fix it later.